container/lru/map.go

```diff
@@ -0,0 +1,524 @@
+// the max limit.
+//
+// The expiration time for existing items is also updated according to the
+// passed TTL.  A TTL of zero disables expiration for the item.
+//
+// It returns the number of evicted items which will either be 0 or 1.
+//
+// Since this method is for internal use only, it MUST only be called with
+// m.limit > 0 for correct behavior.
+//
+// This function MUST be called with the embedded mutex locked (for writes).
+func (m *Map[K, V]) put(key K, value V, now time.Time, ttl time.Duration) (numEvicted uint32) {
+	// Treat zero TTLs as 100 years in the future in order to allow individual
+	// elements to effectively have expiration disabled without needing to store
+	// an additional byte in every element.
+	if ttl == 0 {
+		const oneHundredYears = time.Hour * 24 * 365 * 100
+		ttl = oneHundredYears
+	}
+
+	// When the entry already exists move it to the front of the list thereby
+	// marking it most recently used.
+	if elem, exists := m.items[key]; exists {
+		elem.value = value
+		elem.expiresAfter = now.Add(ttl).UnixNano()
+		m.moveElemToFront(elem)
+		return 0
+	}
+
+	// Evict the least recently used entry (back of the list) if the new entry
+	// would exceed the size limit.  Also reuse the node so a new one doesn't
+	// have to be allocated.
+	if uint32(len(m.items))+1 > m.limit {
+		// Evict least recently used item.
+		elem := m.root.prev
+		delete(m.items, elem.key)
+
+		// Reuse the list element of the item that was just evicted for the new
+		// item.
+		elem.key = key
+		elem.value = value
+		elem.expiresAfter = now.Add(ttl).UnixNano()
+		m.moveElemToFront(elem)
+		m.items[key] = elem
+		return 1
+	}
+
+	// The limit hasn't been reached yet, so just add the new item.  Reuse old
+	// element nodes when possible.
+	var elem *element[K, V]
+	if len(m.elems) > 0 {
+		finalCacheIdx := len(m.elems) - 1
+		elem = m.elems[finalCacheIdx]
+		m.elems = m.elems[0:finalCacheIdx]
+		elem.key = key
+		elem.value = value
+	} else {
+		elem = &element[K, V]{key: key, value: value}
+	}
+	elem.expiresAfter = now.Add(ttl).UnixNano()
+	m.insertElemFront(elem)
+	m.items[key] = elem
+	return 0
+}
+
+// Put either adds the passed key/value pair when an item for that key does not
+// already exist or updates the existing item for the given key to the passed
+// value and arranges for the item to expire after the configured default TTL
+// (if any).  The associated item becomes the most recently used item.
+//
+// The least recently used item is evicted when adding a new item would exceed
+// the max limit.
+//
+// It returns the number of evicted items which includes any items that were
+// evicted due to being marked expired.
+//
+// See [PutWithTTL] for a variant that allows setting a specific TTL for the
+// item.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Put(key K, value V) (numEvicted uint32) {
+	defer m.mtx.Unlock()
+	m.mtx.Lock()
+
+	// Nothing can be added when the limit is zero.
+	if m.limit == 0 {
```

Comment:
Given that a Set or Map with limit set to 0 is effectively useless, would it make sense to move this check to the constructor and return an error if the caller passes 0 as a param?

Reply:
Thanks for the review thus far.

I deleted my original comment since I thought it was referring to somewhere else when I saw the mail and didn't first look at context, so if you got notified, please ignore that.

It's not a bad suggestion and I did consider that, but I reasoned that even though it's useless, 0 is not really an error and I prefer to avoid errors in constructors when possible because they tend to be viral.  What I mean by that is anything that wanted to use an LRU, almost always needs to construct the LRU in their own constructors and is the LRU constructor returned an error, now the  calling constructor has to deal with that error, almost positively by requiring it to also return and error and so on.